### PR TITLE
make it cleaner

### PR DESCRIPTION
**Pull request by remstone7**
time to just refactor this and add all the endpoints because it's nice

### CODE REVIEW

File: klaviyo/api_helper.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            url (str): the url to make the request to
+            params (dict or json): the body of the request
+        Returns:
+            (str, dict): public returns 1 or 0  (pass/fail)
+                        v1/v2 returns dict
+        """
+        self.__is_valid_request_option(request_type=request_type)
+        headers = {
+            'Content-Type': "application/json",
+            'User-Agent': 'Klaviyo/Python/3.0'
```
Comment (klaviyojad) [inline klaviyo/api_helper.py]: make this a constant?

File: README.md [hunk 2]
```diff
@@ -56,81 +56,82 @@ You can then easily use Klaviyo to track events or identify people.  Note, track
 You can get metrics, a timeline of events and export analytics for a metric.  See here for more https://www.klaviyo.com/docs/api/metrics
 
     # return all metrics
-    client.metrics()
-    
-    # you can paginate through using the page offset
-    client.metrics(page=1)
+    client.get_metrics()
```
Comment (smoucka) [inline README.md]: metrics methods can all be changed to client.Metrics.method_name()

File: klaviyo/api_helper.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        """
+        Making sure the appropriate credentials are there
+        Args:
+            request_type (str): the type of method (private/public)
+        """
+        if request_type == self.PUBLIC and not self.public_token:
+            raise KlaviyoException('Public token is not defined')
+
+        if request_type == self.PRIVATE and not self.private_token:
+            raise KlaviyoException('Private token is not defined')
```
[the remark this answers is not in the document]
Comment (remstone7) [reply klaviyo/api_helper.py]: Yea good question, auth error would be something 401/403, could throw a newly created one here

File: klaviyo/lists.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    def update_list_name_by_id(self, list_id, list_name):
+        """
+        This allows you to update a list's name
+        Args:
+            list_id (str)
+            list_name (str):
+        Returns:
+
+        """
+        params = dict({
```
[the remark this answers is not in the document]
Comment (remstone7) [reply klaviyo/lists.py]: Yea no reason, maybe pycharm yelling at me

File: setup.py [hunk 5]
```diff
@@ -1,18 +1,18 @@
 #!/usr/bin/python
 from setuptools import setup, find_packages
-
+from klaviyo import __version__
 setup(
     name = 'klaviyo',
-    version = '2.0.4',
+    version = __version__,
```
[the remark this answers is not in the document]
Comment (remstone7) [reply setup.py]: Its defined in init, imported from there.  Havent built the package fully to test locally but i believe this is okay

**Comment by klaviyojad** [comment]
I m testing it in iPython and getting the errors below

```
In [1]: import klaviyo
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-0d7bdbb0ac97> in <module>()
----> 1 import klaviyo

/Users/jadatoui/python-klaviyo/klaviyo/__init__.py in <module>()
      1 __version__ = '3.0.0'
----> 2 from .api import Klaviyo

/Users/jadatoui/python-klaviyo/klaviyo/api.py in <module>()
      1 import sys
      2
----> 3 from .lists import Lists
      4 from .metrics import Metrics
      5 from .profiles import Profiles

/Users/jadatoui/python-klaviyo/klaviyo/lists.py in <module>()
----> 1 from .api_helper import KlaviyoAPI
      2
      3
      4 class Lists(KlaviyoAPI):
      5     LIST = 'list'

/Users/jadatoui/python-klaviyo/klaviyo/api_helper.py in <module>()
----> 1 from abc import ABC
      2 import base64
      3 import datetime
      4 import json
      5 import time

ImportError: cannot import name ABC
```

**Comment by remstone7** [comment]
i'm not getting that error, please make sure you've pulled it down correctly, and have the correct dependencies

image

File: klaviyo/api_helper.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
+from abc import ABC
```
Comment (klaviyojad) [inline klaviyo/api_helper.py]: ???

File: klaviyo/api_helper.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+import base64
+import datetime
+import json
+import time
+
+import requests
+import simplejson
```
Comment (klaviyojad) [inline klaviyo/api_helper.py]: I get ```ImportError: No module named simplejson```

File: klaviyo/api_helper.py (hunk 6 above)
Comment (remstone7) [reply klaviyo/api_helper.py]: https://docs.python.org/3/library/abc.html

File: klaviyo/api_helper.py (hunk 7 above)
Comment (smoucka) [reply klaviyo/api_helper.py]: make sure you're in an environment that has simplejson installed

File: setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smoucka) [reply setup.py]: ahh, i missed that. the diff didn't load for __init__.py because it was too big

File: klaviyo/api_helper.py (hunk 6 above)
Comment (klaviyojad) [reply klaviyo/api_helper.py]: I dont know why I cannot seem to install this.  I tried `pip install abc` to no avail

Does it matter that my Python version is 2.7.10?

Comment (remstone7) [reply klaviyo/api_helper.py]: looks like it's just py3 i'll fix it

File: klaviyo/lists.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        Returns:
+            (list) of profiles that are subscriped
+        """
+        params = {
+            self.EMAILS: emails
+        }
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.SUBSCRIBE), self.HTTP_GET, params)
+
+    def delete_subscriber_from_list(self, list_id, emails):
```
Comment (jordanallain) [inline klaviyo/lists.py]: should subscriber be plural? delete_subscribers_from_list ?

File: klaviyo/api.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+from .public import Public
+
+
+class Klaviyo(object):
+    def __init__(self, public_token=None, private_token=None):
+        self.public_token = public_token,
+        self.private_token = private_token
+
+    def __getattr__(self, item):
+        return KlaviyoAPIDynamicWrapper(item, self)
```
Comment (remstone7) [inline klaviyo/api.py]: is this right?

File: klaviyo/metrics.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        Return:
+            (dict): with data list of metrics
+        """
+        params = {
+            self.PAGE: page,
+            self.COUNT: count,
+        }
+        return self._v1_request(self.METRICS, self.HTTP_GET, params)
+    
+    def get_metrics_timeline(self, since=None, count=TIMELINE_BATCH_SIZE, sort='desc'):
```
Comment (smoucka) [inline klaviyo/metrics.py]: do we wanna constant desc?

File: klaviyo/api_helper.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        return self.__handle_response(response, request_type)
+
+    def __handle_response(self, response, request_type):
+        """
+        A Helper to either return either a response or a handled exception
+        """
+        if response.status_code == 403:
+            raise KlaviyoAuthenticationError('The api key specified is not valid')
+        elif response.status_code == 429:
+            raise KlaviyoRateLimitException(response.json())
```
Comment (remstone7) [inline klaviyo/api_helper.py]: need to parse this to give a key/value of the rate limit time

File: README.md [hunk 12]
```diff
@@ -56,81 +56,104 @@ You can then easily use Klaviyo to track events or identify people.  Note, track
...
 
     # get all metrics for a profile with the default kwargs
     # to paginate the responses you will get a UUID returned from the response, see here for more information
     # https://www.klaviyo.com/docs/api/people#metrics-timeline
-    client.get_profile_metric_timeline(profile_id, metric_id, since=None, count=100, sort='desc')
+    client.Profiles.get_profile_metrics_timeline_by_id(profile_id, metric_id, since=None, count=100, sort='desc')
+
+## Rate Limiting
+  If a rate limit happens it will throw a klaviyo.exceptions.KlaviyoRateLimitException
+  This will contain a detail key with a string value mentioning the time to back off in
```
Comment (jordanallain) [inline README.md]: is this supposed to say "time to back off in seconds" or is that the end of the sentence?

File: klaviyo/api_helper.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    ######################
+    @staticmethod
+    def _normalize_timestamp(timestamp):
+        if isinstance(timestamp, datetime.datetime):
+            timestamp = time.mktime(timestamp.timetuple())
+        return timestamp
+
+    @staticmethod
+    def _filter_params(params):
+        """ To make sre we're passing in params with values """
```
Comment (jordanallain) [inline klaviyo/api_helper.py]: typo, "sure"

File: klaviyo/api_helper.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+            'User-Agent': 'Klaviyo-Python/{}'.format(__version__)
+        }
+
+        response = getattr(requests, method.lower())(url, headers=headers, data=params)
+
+        return self.__handle_response(response, request_type)
+
+    def __handle_response(self, response, request_type):
+        """
+        A Helper to either return either a response or a handled exception
```
Comment (jordanallain) [inline klaviyo/api_helper.py]: superfluous "either"

File: klaviyo/lists.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+            (dict) containing the list_id
+        """
+        params = {
+            self.LIST_NAME: list_name
+        }
+        return self._v2_request(self.LISTS, self.HTTP_POST, params)
+
+    def get_list_by_id(self, list_id):
+        """
+        This will fetch a list by it's ID
```
Comment (jordanallain) [inline klaviyo/lists.py]: super nitpicking here but it should be "its" with no apostrophe

File: klaviyo/lists.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+        """
+        params = {
+            self.LIST_NAME: list_name
+        }
+
+        return self._v2_request('{}/{}'.format(self.LIST, list_id), self.HTTP_PUT, params)
+        
+    def delete_list(self, list_id):
+        """
+        Deletes a list by it's ID
```
Comment (jordanallain) [inline klaviyo/lists.py]: same nitpick as earlier, should be "its" with no apostrophe

File: klaviyo/lists.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+            self.PROFILES: profiles
+        }
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.SUBSCRIBE), self.HTTP_POST, params)
+
+    def get_list_subscription_status(self, list_id, emails):
+        """
+        Check if profiles are on a list and not suppressed
+        Args:
+            list_id (str): klaviyo list id
+            emails (list): a list of email address
```
Comment (jordanallain) [inline klaviyo/lists.py]: addresses should be plural

File: klaviyo/lists.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+        Returns:
+            (list) of profiles that are subscriped
```
Comment (jordanallain) [inline klaviyo/lists.py]: typo, "subscribed"

File: klaviyo/lists.py [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+        }
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.SUBSCRIBE), self.HTTP_GET, params)
+
+    def delete_subscribers_from_list(self, list_id, emails):
+        """
+        Delete and remove profiles from list
+        Args:
+            list_id (str): klaviyo list id
+            emails (list): a list of email address
```
Comment (jordanallain) [inline klaviyo/lists.py]: "addresses" plural

File: klaviyo/lists.py [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.SUBSCRIBE), self.HTTP_DELETE, params)
+
+    def add_members_to_list(self, list_id, profiles):
+        """
+        This will just add a user to a list regardless of the settings
+        Args:
+            list_id (str): klaviyo list id
+            profiles (dict): for POST -> data must be a list of objects
+        ReturnsL
```
Comment (jordanallain) [inline klaviyo/lists.py]: typo, L instead of :

File: klaviyo/lists.py [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+            self.PROFILES: profiles
+        }
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.MEMBERS), self.HTTP_POST, params)
+
+    def get_list_membership_status(self, list_id, emails):
+        """
+        Check if profiles are on a list
+        Args:
+            list_id (str): klaviyo list id
+            emails (list): a list of email address
```
Comment (jordanallain) [inline klaviyo/lists.py]: plural, "addresses"

File: klaviyo/lists.py [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+        }
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.MEMBERS), self.HTTP_GET, params)
+
+    def delete_list_membership_status(self, list_id, emails):
+        """
+        Remove profiles from a list
+        Args:
+            list_id (str): klaviyo list id
+            emails (list): a list of email address
```
Comment (jordanallain) [inline klaviyo/lists.py]: plural, "addresses"

File: klaviyo/metrics.py [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        params = {
+            self.PAGE: page,
+            self.COUNT: count,
+        }
+        return self._v1_request(self.METRICS, self.HTTP_GET, params)
+    
+    def get_metrics_timeline(self, since=None, count=TIMELINE_BATCH_SIZE, sort=KlaviyoAPI.SORT_DESC):
+        """"
+        Fetches all of the metrics and it's events regardless of the statistic
```
Comment (jordanallain) [inline klaviyo/metrics.py]: should be "its" without apostrophe

File: klaviyo/metrics.py [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+
+        return self._v1_request(url, self.HTTP_GET, params)
+        
+    def get_metric_timeline_by_id(self, metric_id, since=None, count=TIMELINE_BATCH_SIZE, sort=KlaviyoAPI.SORT_DESC):
+        """"
+        Returns a timeline of events for a specific metric
+        Args:
+            metric_id (str): metric ID for the statistic
+            since (str or int): next attribute of the previous api call or unix timestamp
+            count (int): number of events retuned
```
Comment (jordanallain) [inline klaviyo/metrics.py]: typo, "returned"

File: klaviyo/profiles.py [hunk 25]
```diff
@@ -0,0 +1,75 @@
+from .api_helper import KlaviyoAPI
+
+class Profiles(KlaviyoAPI):
+    PERSON = 'person'
+
+    def get_profile(self, profile_id):
+        """
+        https://www.klaviyo.com/docs/api/people#person
+        Get a profile by it's ID
```
Comment (jordanallain) [inline klaviyo/profiles.py]: should be "its" again no apostrophe

File: klaviyo/public.py [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+            properties = {}
+
+        properties[self.TRACK_ONCE_KEY] = True
+
+        return self.track(event, email=email, id=id, properties=properties, customer_properties=customer_properties,
+            ip_address=ip_address, is_test=is_test)
+
+    def identify(self, email=None, id=None, properties={}, is_test=False):
+        """
+        Makes an identify call to Klaviyo API. This will create/update a user with it's associated customer properties
```
Comment (jordanallain) [inline klaviyo/public.py]: "its", no apostrophe

File: klaviyo/api_helper.py [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+
+        # we return the str 1 or 0 for track/identify calls
+        elif request_type == self.PUBLIC:
+            return response.text
+
+    def __handle_200_response(self, response):
+        """
+        A 200 response, either return json or http ok response
+        Args:
+            reponse (obj): Requests object
```
Comment (klaviyojad) [inline klaviyo/api_helper.py]: Response object not request

File: klaviyo/public.py [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+        Args:
+            email (str or None): email address
+            id (str or None): external id for customer
+            properties (dict): information about the customer
+            is_test (bool): should this be a test request
+        Returns:
+            (str): 1 (pass) or 0 (fail)
+        """
+        if email is None and id is None:
```
Comment (klaviyojad) [inline klaviyo/public.py]: if not email and not id:

File: klaviyo/middleware.py [hunk 29]
```diff
@@ -7,16 +7,8 @@ def is_html_response(response):
     return 'text/html' in response.get('Content-Type', '')
 
 def render_script(api_token):
-        return """<script text="text/javascript">
-  var _learnq = _learnq || [];
-  _learnq.push(['account', '{}']);
+        return """<script type="application/javascript" async src="https://static.klaviyo.com/onsite/js/klaviyo.js?company_id={}"></script>""".format(api_token)
```
Comment (jordanallain) [inline klaviyo/middleware.py]: isn't this supposed to be `type="text/javascript"` ?

**Pull request closed without merging**